### PR TITLE
Ensure safe numbers

### LINKED ISSUE #9: Issues regarding pathSets including integer-like values
So i'm currently working on a project where we have defined multiple routes where we do something like {keys:productIds}, we does this because our productIds are string, but now and then we get id's that your isNumber checking function here https://github.com/Netflix/falcor-path-utils/blob/master/lib/toPaths.js#L218 falsely flags as numbers, it is string like this one "0123", which is to say an integer with leading zeros. This results in falcor, routes containing actual integers, routes we haven't defined, and do not intend to, because our id's are strings, so it wouldn't make sense.

I would like to propose a change that takes a more aggressive stance, in order to just enforce the number type, so the check would maybe look a bit more like this:
`typeof value === 'number' && isFinite(val) && val !== Infinity && val !== -Infinity`

### PR DESCRIPTION
**Pull request by lrowe**
Rebase of #6 adding with a tweak make isSafeNumber exclude non integers. Fixes #9.

### CODE REVIEW

File: lib/toPaths.js [hunk 1]
```diff
@@ -1,13 +1,18 @@
 var isArray = Array.isArray;
 var typeOfObject = "object";
+var typeOfString = "string";
+var typeOfNumber = "number";
+var MAX_SAFE_INTEGER = 9007199254740991; // Number.MAX_SAFE_INTEGER in es6
```
Comment (sdesai) [inline lib/toPaths.js:5]: We should probably set this up as a polyfill [ that is, rely on Number.MAX_SAFE_INTEGER if we have it, else fallback to a hardcoded value ]

File: lib/toPaths.js [hunk 2]
```diff
@@ -207,14 +212,35 @@ function getHashCode(key) {
...
-    // adding 1 corrects loss of precision from parseFloat (#15100)
-    return !isArray(val) && (val - parseFloat(val) + 1) >= 0;
+function isSafeNumber(val) {
+    var num = val;
+    var type = typeof val;
+    if (type === typeOfString) {
+        var length = val.length;
+        // Number.MIN_SAFE_INTEGER is 17 digits including the sign.
+        // Anything longer cannot be safe.
+        if (length === 0 || length > 17) {
```
Comment (sdesai) [inline lib/toPaths.js]: Same polyfill note here. Plus probably want to assign it to named variable/constant, like we do for MAX_SAFE_INTEGER [ so we don't have magic numbers floating around ]

File: lib/toPaths.js [hunk 3]
```diff
@@ -207,14 +212,35 @@ function getHashCode(key) {
...
+            return false;
+        }
+        if (!safeNumberRegEx.test(val)) {
```
Comment (sdesai) [inline lib/toPaths.js:231]: Is this why you were trying to get performance numbers?

File: lib/toPaths.js [hunk 4]
```diff
@@ -207,14 +212,35 @@ function getHashCode(key) {
...
+            return false;
+        }
+        // Number.MAX_SAFE_INTEGER is 16 digits.
+        // Anything shorter must be safe.
+        if (length < 16) {
```
Comment (sdesai) [inline lib/toPaths.js]: Same polyfill, magic number note

File: lib/toPaths.js [hunk 5]
```diff
@@ -1,13 +1,18 @@
 var isArray = Array.isArray;
 var typeOfObject = "object";
+var typeOfString = "string";
+var typeOfNumber = "number";
+var MAX_SAFE_INTEGER = 9007199254740991; // Number.MAX_SAFE_INTEGER in es6
+var abs = Math.abs;
+var safeNumberRegEx = /^(0|(\-?[1-9][0-9]*))$/;
```
Comment (sdesai) [inline lib/toPaths.js:9]: I'm just wondering if not treating 03 (for example) as a number is a conscious design decision or rather what the design motivation is for it. I'm not sure if we have it spec'd anywhere, so wondering if it's just a "that's probably what a user would expect" or a "not likely to happen" thing.

File: lib/toPaths.js (hunk 1 above)
Comment (lrowe) [reply lib/toPaths.js:5]: I figured as the number is fixed we may as well use the shorter code here. (Alternative `var MAX_SAFE_INTEGER = Number.MAX_SAFE_INTEGER || 9007199254740991;`)

File: lib/toPaths.js (hunk 5 above)
Comment (lrowe) [reply lib/toPaths.js:9]: Falcor treats numbers and string representations of numbers interchangeably, so we need to make sure that any strings we convert to integers roundtrip convert back to the same string, i.e. the user input `'03'` should not be converted to the string `'3'` by the time it hits the get handler.

File: lib/toPaths.js (hunk 2 above)
Comment (lrowe) [reply lib/toPaths.js]: These constants aren't available anyway, but will move them to a variable.

File: lib/toPaths.js (hunk 3 above)
Comment (lrowe) [reply lib/toPaths.js:231]: Yup. However I'm assured that regex is fast even on slow platforms so we should be ok.

**Pull request merged**